File: implementation/C/lbm_py2c/lbmFlowAroundCylinder.c
/*!
 * \file    lbmFlowAroundCylinder.c
 * \brief   (Almost) "One to one" C implementation of lbmFlowAroundCylinder.py
 * \author  Adrien Python
 * \date    21.11.2016
 */

#define _GNU_SOURCE

#include <stdlib.h>
#include <unistd.h>
#include <stdio.h>
#include <math.h>
#include <stdbool.h>
#include <libgen.h>
#include <pgm.h>
#include <array.h>

#ifndef NX
#define NX       420         // Numer of lattice nodes (width)
#endif
#ifndef NY
#define NY       180         // Numer of lattice nodes (height)
#endif
#define RE       220.0       // Reynolds number
#define LY       ((NY) - 1)  // Height of the domain in lattice units
#define CX       ((NX) / 4)  // X coordinates of the cylinder
#define CY       ((NY) / 2)  // Y coordinates of the cylinder
#define R        ((NY) / 9)  // Cylinder radius
#define ULB      0.04        // Velocity in lattice units
#define NULB     ((ULB) * (R) / (RE))   // Viscoscity in lattice units
#define OMEGA    (1. / (3*(NULB)+0.5))  // Relaxation parameter

#define SQUARE(a) ((a)*(a))

typedef enum { OUT_NONE, OUT_FIN, OUT_IMG } out_mode;

const ssize_t V[][2] = {
    { 0, 0},
    {-1, 1}, {-1, 0}, {-1,-1}, { 0,-1},
    { 1,-1}, { 1, 0}, { 1, 1}, { 0, 1}
};

const double T[] = {
    4./9,
    1./36, 1./9, 1./36, 1./9,
    1./36, 1./9, 1./36, 1./9
};

/**
 * Setup: cylindrical obstacle and velocity inlet with perturbation
 * Creation of a mask with boolean values, defining the shape of the obstacle.
 */
static void initObstacles(bool** obstacles)
{
    for (int x = 0; x < NX; x++) {
        for (int y = 0; y < NY; y++) {
            obstacles[x][y] = SQUARE(x-CX) + SQUARE(y-CY) < SQUARE(R);
        }
    }
}

/**
 * Initial velocity profile: almost zero, with a slight perturbation to trigger
 * the instability.
 */
static void initVelocity(double*** vel)
{
    for (int d = 0; d < 2; d++) {
        for (int x = 0; x < NX; x++) {
            for (int y = 0; y < NY; y++) {
                vel[d][x][y] = (1-d) * ULB * (1 + 0.0001 * sin( y / (double)LY * 2 * M_PI) );
            }
        }
    }
}

/**
 * Equilibrium distribution function.
 */
static void equilibrium(double*** feq, double** rho, double*** u)
{
    double usqr[NX][NY];
    
    for (int x = 0; x < NX; x++) {
        for (int y = 0; y < NY; y++) {
            usqr[x][y] = 3./2 * ( SQUARE(u[0][x][y]) + SQUARE(u[1][x][y]) );
        }
    }
    
    double cu[NX][NY];
    
    for (int f = 0; f < 9; f++) {
        for (int x = 0; x < NX; x++) {
            for (int y = 0; y < NY; y++) {
                cu[x][y] = 3 * ( V[f][0] * u[0][x][y] + V[f][1] * u[1][x][y] );
                feq[f][x][y] = rho[x][y] * T[f] * ( 1 + cu[x][y] + 0.5 * SQUARE(cu[x][y]) - usqr[x][y] );
            }
        }
    }
}

static void macroscopic(double*** fin, double** rho, double*** u)
{
    // density (rho) is the sum of the nine populations
    for (int x = 0; x < NX; x++) {
        for (int y = 0; y < NY; y++) {
            rho[x][y] = 0;
            for (int f = 0; f < 9; f++) {
                rho[x][y] += fin[f][x][y];
            }
        }
    }
    
    {
        double zero = 0;
        array_set3(2, NX, NY, sizeof(double), u, &zero);
    }
    
    for (int f = 0; f < 9; f++) {
        for (int x = 0; x < NX; x++) {
            for (int y = 0; y < NY; y++) {
                u[0][x][y] += V[f][0] * fin[f][x][y];
                u[1][x][y] += V[f][1] * fin[f][x][y];
            }
        }
    }
    
    for (int d = 0; d < 2; d++) {
        for (int x = 0; x < NX; x++) {
            for (int y = 0; y < NY; y++) {
                u[d][x][y] /= rho[x][y];
            }
        }
    }
}

static void initCol(size_t* col, ssize_t v0)
{
    for (int f = 0, i = 0; f < 9 && i < 3; f++) {
        if (V[f][0] == v0) {
            col[i++] = f;
        }
    }
}

static void initOpp(size_t* opp)
{
    for (int f = 0; f < 9; f++) {
        for (int g = 0; g < 9; g++) {
            if (V[f][0] == -V[g][0] && V[f][1] == -V[g][1]) {
                opp[f] = g;
                break;
            }
        }
    }
}

static bool write_double(size_t dim, void* data, const size_t index[dim], void* args)
{
    (void) dim, (void) index;

    fprintf((FILE*)args, "%64.60f\n", *((double*)data));
    
    return true;
}

static inline void output_variables(char* filename, size_t dim, const size_t index[dim], void* array)
{
    FILE* file = fopen(filename, "w");
    array_foreach(dim, index, sizeof(double), array, NULL, write_double, file);
    fclose(file);
}

void output_image(char* filename, double*** u) 
{
    pgm_image* pgm = pgm_create(NX, NY);
    
    for (size_t x = 0; x < NX; x++) {
        for (size_t y = 0; y < NY; y++) {
            double vel = sqrt( SQUARE(u[0][x][y]) + SQUARE(u[1][x][y]) );
            int color =  255 * vel * 10;
            pgm_set_pixel(pgm, x, y, color);
        }
    }

    pgm_write(pgm, filename);
    pgm_destroy(pgm);
}

int main(int argc, char * const argv[])
{
    // Init options to default values
    const char* out_path = ".";
    const char* out_pref = "lbm";
    out_mode out = OUT_NONE;
    ssize_t max_iter = 0;
    size_t out_interval = 0;

    // Read arguments
    while (optind < argc) {
        switch (getopt(argc, argv, "pfi:I:o:O:")) {
            case 'p': { out = OUT_IMG; break; }
            case 'f': { out = OUT_FIN; break; }
            case 'i': { max_iter = strtol(optarg, NULL, 10); break; }
            case 'I': { out_interval = strtol(optarg, NULL, 10); break; }
            case 'o': { out_path = optarg; break; }
            case 'O': { out_pref = optarg; break; }
            default : { goto usage; }
        }
    }
    
    // check that execution mode is set (output images or fin values)
    if (max_iter < 1) {
    usage:
        fprintf(stderr, "usage: %s (-p | -f) -i <iter> [-I <out_interval>] [-o <out_dir>] [-O <out_prefix>]\n", basename((char*)argv[0]));
        fprintf(stderr, "  -p : output pictures\n");
        fprintf(stderr, "  -f : output populations\n");
        fprintf(stderr, "  -i : number of iterations\n");
        fprintf(stderr, "  -I : output interval; (0 if only the last iteration output in required)\n");
        fprintf(stderr, "  -o : output file directory\n");
        fprintf(stderr, "  -O : output filename prefix\n");
        return EXIT_FAILURE;
    }

    if (out == OUT_NONE) {
        fprintf(stderr, "No output mode specified.\n");
    }
    
    const size_t A_SIZE0[3] = {2, NX, NY};
    const size_t A_SIZE1[3] = {9, NX, NY};
    const size_t A_SIZE2[2] = {NX, NY};
    
    double*** u = array_create(3, A_SIZE0, sizeof(double));    // velocity to impose on inflow boundary cells
    double*** feq = array_create(3, A_SIZE1, sizeof(double));  //
    double*** fin = array_create(3, A_SIZE1, sizeof(double));  // Incoming populations
    double*** fout = array_create(3, A_SIZE1, sizeof(double)); // Outgoing populations
    double** rho = array_create(2, A_SIZE2, sizeof(double));   // Density
    bool** obstacles = array_create(2, A_SIZE2, sizeof(bool)); // Obstacles definition
    double*** vel = array_create(3, A_SIZE0, sizeof(double));  // Velocity

    size_t COL1[3], COL2[3], COL3[3], OPP[9];

    initCol(COL1,  1);
    initCol(COL2,  0);
    initCol(COL3, -1);
    initOpp(OPP);

    (void) COL1;
    
    {
        double one = 1;
        array_set(2, A_SIZE2, sizeof(double), rho, &one);
    }
    
    initObstacles(obstacles);
    initVelocity(vel);
    
    // Initialization of the populations at equilibrium with the given velocity.
    equilibrium(fin, rho, vel);
        
    for (int iter = 1; iter <= max_iter; iter++) {
        
        // Right wall: outflow condition.
        for (int i = 0; i < 3; i++) {
            for (size_t y = 0, f = COL3[i]; y < NY; y++) {
                fin[f][NX-1][y] = fin[f][NX-2][y];
            }
        }
        
        // Compute macroscopic variables, density and velocity
        macroscopic(fin, rho, u);
        
        // Left wall: inflow condition
        for (size_t d = 0; d < 2; d++) {
            for (size_t y = 0; y < NY; y++) {
                u[d][0][y] = vel[d][0][y];
            }
        }
        
        // Calculate the density
        for (size_t y = 0; y < NY; y++) {
            double s2 = 0, s3 = 0;
            for (size_t i = 0; i < 3; i++) {
                s2 += fin[COL2[i]][0][y];
                s3 += fin[COL3[i]][0][y];
            }
            rho[0][y] = 1./(1 - u[0][0][y]) * (s2 + 2*s3);
        }
        
        // Compute equilibrium
        equilibrium(feq, rho, u);
        for (size_t i = 0; i < 3; i++) {
            for (size_t y = 0, f = COL1[i]; y < NY; y++) {
                fin[f][0][y] = feq[f][0][y] + fin[OPP[f]][0][y] - feq[OPP[f]][0][y];
            }
        }
     
        // Collision step
        for (size_t f = 0; f < 9; f++) {
            for (size_t x = 0; x < NX; x++) {
                for (size_t y = 0; y < NY; y++) {
                    fout[f][x][y] = fin[f][x][y] - OMEGA * (fin[f][x][y] - feq[f][x][y]);
                }
            }
        }
        
        // Bounce-back condition for obstacle
        for (size_t f = 0; f < 9; f++) {
            for (size_t x = 0; x < NX; x++) {
                for (size_t y = 0; y < NY; y++) {
                    if (obstacles[x][y]) {
                        fout[f][x][y] = fin[OPP[f]][x][y];
                    }
                }
            }
        }
        
        // Streaming step
        for (size_t f = 0; f < 9; f++) {
            array_roll2_to(NX, NY, sizeof(double), fout[f], fin[f], V[f]);
        }
        
        if ( (!out_interval && iter == max_iter) || (out_interval && iter % out_interval == 0) ) {

            char* filename;

            if ( out == OUT_IMG ) {
                if ( asprintf(&filename, "%s/%s%d.pgm", out_path, out_pref, iter) != -1 ) {
                    output_image(filename, u);
                    free(filename);
                }
            }

            if (out == OUT_FIN) {
                if ( asprintf(&filename, "%s/%s%d.out", out_path, out_pref, iter) != -1 ) {
                    output_variables(filename, 3, A_SIZE1, fin);
                    free(filename);
                }
            }
        }
    }
    
    array_destroy(3, A_SIZE0, sizeof(double), u);
    array_destroy(3, A_SIZE1, sizeof(double), feq);
    array_destroy(3, A_SIZE1, sizeof(double), fin);
    array_destroy(3, A_SIZE1, sizeof(double), fout);
    array_destroy(2, A_SIZE2, sizeof(double), rho);
    array_destroy(2, A_SIZE2, sizeof(bool), obstacles);
    array_destroy(3, A_SIZE0, sizeof(double), vel);
    
    return EXIT_SUCCESS;
}

File: implementation/Python/lbm_simple_3d/lbmFlowAroundCylinder.py
#!/usr/bin/env python3
# Copyright (C) 2015 Universite de Geneve, Switzerland
# E-mail contact: jonas.latt@unige.ch
#
# 2D flow around a cylinder
#
# Update by Adrien Python (adrien.python@gmail.com):
# Slightly modified from original to handle different ordering for v/t variables

from numpy import *
import matplotlib.pyplot as plt
from matplotlib import cm
import sys
import datetime

###### Outputs definitions #####################################################
out = sys.argv[1]
outInterval = int(sys.argv[3])
outDir  = sys.argv[4]
outPre  = sys.argv[5]

###### Flow definition #########################################################
maxIter = int(sys.argv[2])  # Total number of time iterations.
Re = 220.0         # Reynolds number.
nx, ny, nz = 100, 100, 10 # Numer of lattice nodes.
ly = ny-1         # Height of the domain in lattice units.
cx, cy, cz, r = nx//4, ny//2, nz//2, ny//9 # Coordinates of the cylinder.
uLB     = 0.04                  # Velocity in lattice units.
nulb    = uLB*10/Re;             # Viscoscity in lattice units.
omega = 1 / (3*nulb+0.5);    # Relaxation parameter.

###### Lattice Constants #######################################################


v = array([[ 1,  1,  0], [ 1,  0,  0], [ 1, -1,  0], [ 0,  1,  0], [  0,  0,  0], [ 0, -1, 0], [-1,  1, 0], [-1,  0, 0], [-1, -1, 0],
           [ 1,  0,  1], [ 0,  1,  1], [ 0,  0,  1], [ 0, -1,  1], [ -1,  0,  1], 
           [ 1,  0, -1], [ 0,  1, -1], [ 0,  0, -1], [ 0, -1, -1], [ -1,  0, -1]])
t = array([  1./36, 1./18, 1./36, 1./18, 1./3 , 1./18, 1./36, 1./18, 1./36, 
             1./36, 1./36, 1./18, 1./36, 1./36, 
             1./36, 1./36, 1./18, 1./36, 1./36]) 

###### Automaticaly defined Constants ##########################################

def initcol(v0):
    return [f for f in range(19) if v[f][0] == v0]

col1 = initcol(1)
col2 = initcol(0)
col3 = initcol(-1)

def initopp():
    opp = [None] * 19
    for f in range(19):
        for g in range(19):
            if v[f][0] == -v[g][0] and v[f][1] == -v[g][1] and v[f][2] == -v[g][2]:
                opp[f] = g
                break
    return opp

opp  = initopp();

###### Function Definitions ####################################################
def macroscopic(fin):
    rho = sum(fin, axis=0)
    u = zeros((3, nx, ny, nz))
    for i in range(19):
        u[0,:,:,:] += v[i,0] * fin[i,:,:,:]
        u[1,:,:,:] += v[i,1] * fin[i,:,:,:]
        u[2,:,:,:] += v[i,2] * fin[i,:,:,:]
    u /= rho
    return rho, u

def equilibrium(rho, u):              # Equilibrium distribution function.
    usqr = 3/2 * (u[0]**2 + u[1]**2 + u[2]**2)
    feq = zeros((19,nx,ny,nz))
    for i in range(19):
        cu = 3 * (v[i,0]*u[0,:,:,:] + v[i,1]*u[1,:,:,:] + v[i,2]*u[2,:,:,:])
        feq[i,:,:,:] = rho*t[i] * (1 + cu + 0.5*cu**2 - usqr)
    return feq

###### Setup: cylindrical obstacle and velocity inlet with perturbation ########
# Creation of a mask with 1/0 values, defining the shape of the obstacle.
def obstacle_fun(x, y, z):
    return (x-cx)**2+(y-cy)**2+(z-cz)**2<0 #(x-cx)**2+(y-cy)**2<r**2

obstacle = fromfunction(obstacle_fun, (nx,ny,nz))

# Initial velocity profile: almost zero, with a slight perturbation to trigger
# the instability.
def inivel(d, x, y, z):
    return 0*y #(1-d) * uLB * (1 + 1e-4*sin(y/ly*2*pi))

vel = fromfunction(inivel, (3,nx,ny,nz))

# Initialization of the populations at equilibrium with the given velocity.
rho = ones((nx,ny,nz))
rho[nx//2,ny//2,nz//2] = 2
fin = equilibrium(rho, vel)

seconds = 0
start = datetime.datetime.now()

if out == 'OUT':
    file = open("{0}/{1}{2}.out".format(outDir, outPre, 0), 'w')
    for x in range(nx):
        for y in range(ny):
            for z in range(nz):
                for f in range(19):
                    file.write("{0:64.60f}\n".format(fin[f,x,y,z]))


###### Main time loop ##########################################################
for iter in range(1, maxIter+1):

    # Compute macroscopic variables, density and velocity.
    rho, u = macroscopic(fin)

    # Compute equilibrium.
    feq = equilibrium(rho, u)

    # Collision step.
    fout = fin - omega * (fin - feq)

    # Bounce-back condition for obstacles
    for i in range(19):
        fout[i, obstacle] = fin[opp[i], obstacle]

    # Streaming step.
    for i in range(19):
        fin[i,:,:,:] =  roll(
                            roll(
                                roll(fout[i,:,:,:], v[i,0], axis=0),
                                v[i,1], axis=1 ),
                            v[i,2], axis=2 )

    if (outInterval==0 and iter == maxIter) or (outInterval>0 and iter % outInterval == 0):
        
        seconds += (datetime.datetime.now() - start).total_seconds()

        # Visualization of the velocity.
        if out == 'IMG':
            unorm_box = sqrt(u[0]**2+u[1]**2+u[2]**2).transpose()
            plt.clf()
            plt.imshow(unorm_box[nz//2,:,:], cmap=cm.Reds)
            plt.savefig("{0}/{1}{2}.png".format(outDir, outPre, iter))
        if out == 'OUT':
            file = open("{0}/{1}{2}.out".format(outDir, outPre, iter), 'w')
            for x in range(nx):
                for y in range(ny):
                    for z in range(nz):
                        for f in range(19):
                            file.write("{0:64.60f}\n".format(fin[f,x,y,z]))
        start = datetime.datetime.now()

print ("average lups:", nx*ny*maxIter/max(1,seconds))
